File: Level 2/Level_2_다리를 지나는 트럭.cpp
/* 프로그래머스 - Level 2 다리를 지나는 트럭 */

#include <queue>
#include <vector>

using namespace std;

struct Truck {
	int loaded_second;
	int weight;

	Truck(int loaded_second, int weight) : loaded_second(loaded_second), weight(weight) {}
};

int solution(int bridge_length, int weight, vector<int> truck_weights) {
	queue<Truck> bridge;

	int second = 0, loaded_weight = 0, waiting = 0;
	while (waiting < truck_weights.size()) {
		second += 1;

		if (!bridge.empty()) {
			Truck front_truck = bridge.front();
			if (second - front_truck.loaded_second == bridge_length) {
				bridge.pop();
				loaded_weight -= front_truck.weight;
			}
		}

		int truck_weight = truck_weights[waiting];
		if (loaded_weight + truck_weight <= weight) {
			bridge.push(Truck(second, truck_weight));
			loaded_weight += truck_weight;
			waiting += 1;
		}
	}
	second += bridge_length;
	
	return second;
}

File: Level 2/Level_2_다리를 지나는 트럭.kt
/* 프로그래머스 - Level 2 다리를 지나는 트럭 */

package level2

import java.util.*

class Solution {
    data class Truck(val loadedSecond: Int, val weight: Int)

    fun solution(bridge_length: Int, weight: Int, truck_weights: IntArray): Int {
        var bridge: Queue<Truck> = LinkedList<Truck>()

        var second = 0
        var loadedWeight = 0
        var waiting = 0
        while(waiting < truck_weights.size) {
            second++

            if(!bridge.isEmpty()) {
                val frontTruck = bridge.peek()
                if(second - frontTruck.loadedSecond == bridge_length) {
                    bridge.poll()
                    loadedWeight -= frontTruck.weight
                }
            }

            val truckWeight = truck_weights[waiting]
            if(loadedWeight + truckWeight <= weight) {
                bridge.add(Truck(second, truckWeight))
                loadedWeight += truckWeight
                waiting += 1
            }
        }
        second += bridge_length

        return second
    }
}